collector/os_release.go

```diff
@@ -0,0 +1,172 @@
+}
+
+type osReleaseCollector struct {
+	infoDesc    *prometheus.Desc
+	logger      log.Logger
+	os          *osRelease
+	osFilename  string
+	osMtime     time.Time
+	osMutex     sync.Mutex
+	version     float64
+	versionDesc *prometheus.Desc
+}
+
+func init() {
+	registerCollector("os", defaultEnabled, NewOSCollector)
+}
+
+// NewOSCollector returns a new Collector exposing os-release information.
+func NewOSCollector(logger log.Logger) (Collector, error) {
+	return &osReleaseCollector{
+		logger: logger,
+		infoDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "info"),
+			"A metric with a constant '1' value labeled by build_id, id, id_like, image_id, image_version, "+
+				"name, pretty_name, variant, variant_id, version, version_codename, version_id.",
+			[]string{"build_id", "id", "id_like", "image_id", "image_version", "name", "pretty_name",
+				"variant", "variant_id", "version", "version_codename", "version_id"}, nil,
+		),
+		versionDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "os", "version"),
+			"Metric containing the major.minor part of the OS version.",
+			[]string{"id", "id_like", "name"}, nil,
+		),
+	}, nil
+}
+
+func parseOSRelease(r io.Reader) (*osRelease, error) {
+	env, err := envparse.Parse(r)
+	return &osRelease{
+		Name:            env["NAME"],
+		ID:              env["ID"],
+		IDLike:          env["ID_LIKE"],
+		PrettyName:      env["PRETTY_NAME"],
+		Variant:         env["VARIANT"],
+		VariantID:       env["VARIANT_ID"],
+		Version:         env["VERSION"],
+		VersionID:       env["VERSION_ID"],
+		VersionCodename: env["VERSION_CODENAME"],
+		BuildID:         env["BUILD_ID"],
+		ImageID:         env["IMAGE_ID"],
+		ImageVersion:    env["IMAGE_VERSION"],
+	}, err
+}
+
+func (c *osReleaseCollector) UpdateStruct(path string) error {
+	releaseFile, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer releaseFile.Close()
+
+	stat, err := releaseFile.Stat()
+	if err != nil {
+		return err
+	}
+
+	t := stat.ModTime()
+	if path == c.osFilename && t == c.osMtime {
+		// osReleaseCollector struct is already up-to-date.
+		return nil
+	}
+
+	// Acquire a lock to update the osReleaseCollector struct.
+	c.osMutex.Lock()
+	defer c.osMutex.Unlock()
+
+	level.Debug(c.logger).Log("msg", "file modification time has changed",
+		"file", path, "old_value", c.osMtime, "new_value", t)
+	c.osFilename = path
+	c.osMtime = t
+
+	c.os, err = parseOSRelease(releaseFile)
+	if err != nil {
+		return err
+	}
+
+	majorMinor := versionRegex.FindString(c.os.VersionID)
+	if majorMinor != "" {
+		c.version, err = strconv.ParseFloat(majorMinor, 64)
+		if err != nil {
+			return err
+		}
+	} else {
+		c.version = 0
+	}
+	return nil
+}
+
+func (c *osReleaseCollector) Update(ch chan<- prometheus.Metric) error {
+	for i, path := range []string{etcOSRelease, usrLibOSRelease} {
+		err := c.UpdateStruct(*rootfsPath + path)
+		if err == nil {
+			break
+		}
+		if errors.Is(err, os.ErrNotExist) {
+			if i == 1 {
```

Comment:
This is slightly fragile, how about this.

Add a `osReleaseFilenames []string` to `osReleaseCollector struct`.

Then populate `osReleaseFilenames: []string{etcOSRelease, usrLibOSRelease}` in `NewOSCollector()`

Then `range c.osReleaseFilenames` and `if i >= (len(c.osReleaseFilenames) - 1)`.

Lastly, update the debug message to be `("msg", "no os-release files found", "files", strings.Join(c.osReleaseFilenames, ","))`.

This way changing the parsed file list is easier to change if necessary.

Reply:
You are right. Implemented. Also added some comments to the osReleaseCollector struct.